File: skills/jazzqi/xiaomi-mimo-tts/scripts/base/mimo_tts.py
#!/usr/bin/env python3
"""Python base implementation for xiaomi-mimo-tts with dry-run and robust error handling
Usage: python3 mimo_tts.py "文本" [output.ogg] [--voice VOICE] [--style STYLE] [--dry-run]
"""
import sys,os,subprocess, json, base64, urllib.request
if len(sys.argv)<2:
    print('Usage: python3 mimo_tts.py "TEXT" [OUTPUT] [--voice VOICE] [--style STYLE] [--dry-run]')
    sys.exit(2)
text=sys.argv[1]
output=sys.argv[2] if len(sys.argv)>2 and not sys.argv[2].startswith('--') else os.path.join(os.getcwd(),'output.mock.ogg')
voice='mimo_default'
style=''
args=sys.argv[2:]
for i,a in enumerate(args):
    if a=='--voice' and i+1<len(args): voice=args[i+1]
    if a=='--style' and i+1<len(args): style=args[i+1]
DRY = '--dry-run' in sys.argv

XIAOMI_API_KEY=os.environ.get('XIAOMI_API_KEY') or os.environ.get('MIMO_API_KEY')
MOCK = not XIAOMI_API_KEY
if DRY:
    body = {
        "model": "mimo-v2-tts",
        "messages": [
            {"role": "user", "content": "请朗读"},
            {"role": "assistant", "content": text}
        ],
        "audio": {"format": "wav", "voice": voice}
    }
    print('DRY RUN: request payload preview:')
    print(json.dumps(body, ensure_ascii=False, indent=2))
    sys.exit(0)

if MOCK:
    # generate mock silent ogg using ffmpeg if available
    try:
        subprocess.run(['ffmpeg','-f','lavfi','-i','anullsrc=r=16000:cl=mono','-t','0.5','-q:a','9','-acodec','libopus',output,'-y'],check=True,stdout=subprocess.DEVNULL,stderr=subprocess.DEVNULL)
    except Exception:
        open(output,'wb').close()
    print(output)
    sys.exit(0)

# Real implementation: call Xiaomi MiMo API and decode returned base64 audio
body = {
    "model": "mimo-v2-tts",
    "messages": [
        {"role": "user", "content": "请朗读"},
        {"role": "assistant", "content": text}
    ],
    "audio": {"format": "wav", "voice": voice}
}

req = urllib.request.Request(
    'https://api.xiaomimimo.com/v1/chat/completions',
    data=json.dumps(body).encode('utf-8'),
    headers={
        'Authorization': f'Bearer {XIAOMI_API_KEY}',
        'Content-Type': 'application/json'
    }
)
try:
    with urllib.request.urlopen(req, timeout=60) as resp:
        status = getattr(resp, 'status', None)
        resp_text = resp.read().decode('utf-8')
except Exception as e:
    print('API request failed:', e)
    sys.exit(1)

try:
    j = json.loads(resp_text)
except Exception as e:
    print('Invalid JSON response from API')
    print(resp_text)
    sys.exit(1)
if not status or status<200 or status>=300:
    print(f'API returned status {status}')
    if isinstance(j, dict) and j.get('error'):
        print('API error:', j['error'])
    sys.exit(1)

try:
    audio_b64 = j.get('choices', [])[0].get('message', {}).get('audio', {}).get('data')
    if not audio_b64:
        raise ValueError('no audio in response')
except Exception as e:
    print('Failed to parse audio from response:', e)
    print(resp_text)
    sys.exit(1)

wav = base64.b64decode(audio_b64)
wav_path = output + '.wav'
with open(wav_path, 'wb') as f:
    f.write(wav)

# convert wav to ogg if ffmpeg exists
try:
    subprocess.run(['ffmpeg','-y','-i',wav_path,'-acodec','libopus','-b:a','128k',output], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    os.remove(wav_path)
except Exception as e:
    print('ffmpeg conversion failed:', e)
    print('Leaving wav at', wav_path)
    # keep wav as output
    if not os.path.exists(output):
        os.rename(wav_path, output)

print(output)
sys.exit(0)

File: skills/jazzqi/xiaomi-mimo-tts/scripts/base/mimo_tts.js
#!/usr/bin/env node
// Minimal Node.js base implementation for xiaomi-mimo-tts
// Usage: node mimo_tts.js "文本" [output.ogg] [--voice voice] [--style style]

const fs = require('fs');
const { spawnSync } = require('child_process');

const args = process.argv.slice(2);
if (!args[0]) {
  console.error('Usage: node mimo_tts.js "TEXT" [OUTPUT] [--voice VOICE] [--style STYLE]');
  process.exit(2);
}
let text = args[0];
let output = args[1] && !args[1].startsWith('--') ? args[1] : `${process.cwd()}/output.mock.ogg`;
let voice = 'mimo_default';
let style = '';
for (let i=2;i<args.length;i++){
  if (args[i]==='--voice') voice = args[i+1]||voice, i++;
  if (args[i]==='--style') style = args[i+1]||style, i++;
}

const XIAOMI_API_KEY = process.env.XIAOMI_API_KEY || process.env.MIMO_API_KEY || '';
const MOCK = !XIAOMI_API_KEY;

if (DRY) {
  console.log('DRY RUN: request payload preview:');
  console.log(body);
  process.exit(0);
}

if (MOCK) {
  // generate mock silent ogg using ffmpeg if available
  const ff = spawnSync('ffmpeg',['-f','lavfi','-i','anullsrc=r=16000:cl=mono','-t','0.5','-q:a','9','-acodec','libopus',output,'-y']);
  if (ff.error) fs.writeFileSync(output,'');
  console.log(output);
  process.exit(0);
}

// Real implementation: call Xiaomi MiMo API and decode returned base64 audio
const https = require('https');

const body = JSON.stringify({
  model: 'mimo-v2-tts',
  messages: [
    {role: 'user', content: '请朗读'},
    {role: 'assistant', content: text}
  ],
  audio: {format: 'wav', voice}
});

const options = {
  hostname: 'api.xiaomimimo.com',
  port: 443,
  path: '/v1/chat/completions',
  method: 'POST',
  headers: {
    'Authorization': `Bearer ${XIAOMI_API_KEY}`,
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(body)
  }
};

const req = https.request(options, (res) => {
  let data = '';
  res.on('data', (chunk) => data += chunk);
  res.on('end', () => {
    let j;
    try { j = JSON.parse(data); } catch (e) {
      console.error('Invalid JSON response from API');
      console.error(data);
      process.exit(1);
    }
    if (!res.statusCode || res.statusCode < 200 || res.statusCode>=300) {
      console.error(`API returned status ${res.statusCode}`);
      if (j.error) console.error('API error:', j.error);
      process.exit(1);
    }
    try {
      const audio_b64 = j.choices && j.choices[0] && j.choices[0].message && j.choices[0].message.audio && j.choices[0].message.audio.data;
      if (!audio_b64) throw new Error('no audio in response');
      const wav = Buffer.from(audio_b64, 'base64');
      const wavPath = output + '.wav';
      fs.writeFileSync(wavPath, wav);
      // convert to ogg
      const ff = spawnSync('ffmpeg', ['-y','-i',wavPath,'-acodec','libopus','-b:a','128k',output]);
      if (ff.error) { console.error('ffmpeg failed to convert audio:', ff.error); console.error('Leaving wav at', wavPath); process.exit(1); }
      try{ fs.unlinkSync(wavPath);}catch(e){}
      console.log(output);
    } catch (e) {
      console.error('Failed to parse response or extract audio', e.message || e);
      process.exit(1);
    }
  });
});

req.on('error', (e) => { console.error('Request error', e); process.exit(1); });
req.write(body);
req.end();
